## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [ ]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [205]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [206]:
# Use the describe function to gather some basic statistics
df.describe()

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [207]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df[["category", "subcategory", "qty"]]
df

In [208]:
# What three item categories had the most entries?
top_categorie_items = df["category"].value_counts()
top_categories = top_categorie_items.head(3)
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [209]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top_categorie_items.index[0]
top_subcategory = df[df["category"] == top_category]["subcategory"].value_counts().idxmax()
top_subcategory

consumables


In [210]:
# Which five clients had the most entries in the data?
top_client_count = df["client_id"].value_counts()
top_clients = top_client_count.head(5)
top_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [211]:
# Store the client ids of those top 5 clients in a list.
top_client = df['client_id'].value_counts().head(5)
top_clients_list = top_clients.index.to_list()
top_clients_list


[33615, 66037, 46820, 38378, 24741]


In [212]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_clients.index[0]
total_units_client = df[df["client_id"] == top_client_id]["qty"].sum()
total_units_client

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [213]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df["unit_price"] * df["qty"]
df.head()

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [214]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df["unit_weight"].apply(lambda p: 7 * p if p > 50 
                                               else 10 * p)
df.head()


In [215]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = (df["subtotal"] + df["shipping_price"]) * 1.0925
df.head()

In [216]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["cost"] = (df["unit_cost"] * df["qty"]) + df["shipping_price"]
df.head()

In [217]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df["total_price"] - df["cost"]
df.head()

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105   

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [218]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
order_totals = df[df["order_id"].isin(order_ids)].groupby("order_id")["total_price"].sum()
order_totals

Order ID 2742071: Expected Total = 152811.89, Actual Total = 139844.51202499997
Order ID 2173913: Expected Total = 162388.71, Actual Total = 119838.35704999999
Order ID 6128929: Expected Total = 923441.25, Actual Total = 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [219]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_spending = df[df["client_id"].isin(top_clients_list)].groupby("client_id")["total_price"].sum()
client_spending

Client ID 33615: Total Revenue = 6392918.349675
Client ID 66037: Total Revenue = 8746255.3054
Client ID 46820: Total Revenue = 8006332.531775
Client ID 38378: Total Revenue = 9172866.143875003
Client ID 24741: Total Revenue = 76678479.053625


In [220]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = df[df['client_id'].isin(top_clients_list)].groupby('client_id').agg({
    "qty": "sum",
    "shipping_price": "sum",
    "total_price": "sum",
    "profit": "sum"}).reset_index()
summary_df

KeyError: "Column(s) ['Quantity', 'Shipping Price (Millions)'] do not exist"

In [221]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns.
money_column = ["shipping_price", "total_price", "profit"]
summary_df[money_column] = summary_df[money_column].apply(pd.to_numeric, errors="coerce")

# Define a function that converts a dollar amount to millions.
def convert_millions(m):
    return f"${m / 1e6:.2f}M"

# Apply the currency_format_millions function to only the money columns. 
for col in money_column:
    summary_df[col] = summary_df[col].apply(lambda x: convert_millions(x) if pd.notnull(x) else x)
    
# Rename the columns to reflect the change in the money format. 
summary_df = summary_df.rename(columns={
    "qty" : "Total_units", 
    "shipping_price" : "Total_shipping (Millions)",
    "total_price" : "Total_revenue (Millions)",
    "profit" : "Total_profit (Millions)",
})

summary_df

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... Unit Cost Unit Weight Quantity  line_number  \
0          17        2023  ...    762.71        7.50    

In [222]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df = summary_df.sort_values(by="Total_profit (Millions)", ascending=False)
summary_df

In [223]:
#summary - In this assignment, we had a bunch of data and had to organize it neatly using a program called JupyterLab. We did things like adding up how much stuff different clients bought, how much they paid for shipping, and how much money they made overall. Sometimes we ran into problems, but by figuring out what went wrong and fixing it, we learned more about how to work with data using Python.

